File: go/codeinsight/report.go
package codeinsight

// Report is the single self-describing result envelope (contract §3). Every numeric datum is
// reproducible from (Repository, Window) so a Report replays and diffs across history. It is both
// the in-process Go API and the JSON wire shape an edenhttp consumer serves to @eden/visualization.
type Report struct {
	SchemaVersion string        `json:"schemaVersion"`
	Repository    RepositoryRef `json:"repository"`
	Window        Window        `json:"window"`
	Entities      []Entity      `json:"entities"`
	Couplings     []Coupling    `json:"couplings"`
	Ownership     []Ownership   `json:"ownership"`
	Summary       Summary       `json:"summary"`
	Trends        []TrendSeries `json:"trends"`
	Views         []View        `json:"views"`
}

// RepositoryRef identifies the analyzed repository at the commit the snapshot was taken. The
// Identifier is a logical name, never a secret or a URL carrying credentials (contract §3).
type RepositoryRef struct {
	Identifier  string `json:"identifier"`
	HeadCommit  string `json:"headCommit"`
	AnalyzedAt  string `json:"analyzedAt"`
	CommitCount int    `json:"commitCount"`
}

// Window is the slice of history that fed the temporal metrics. The commit hashes make every
// view reproducible at a point in history and let two reports diff cleanly (contract §3).
type Window struct {
	FromCommit string `json:"fromCommit,omitempty"`
	ToCommit   string `json:"toCommit"`
	Since      string `json:"since,omitempty"`
	Until      string `json:"until,omitempty"`
	Revisions  int    `json:"revisions"`
}

// Entity is a per-file node carrying its static + behavioral metrics. Maintainability is the
// LOC-weighted average of the per-function maintainability index (contract §4 / the design fix):
// computing it per function and aggregating keeps a large file from saturating to a uniform E.
type Entity struct {
	Path            string   `json:"path"`
	Kind            string   `json:"kind"`
	Language        string   `json:"language,omitempty"`
	Lines           int      `json:"lines"`
	Cyclomatic      int      `json:"cyclomatic,omitempty"`
	Cognitive       int      `json:"cognitive,omitempty"`
	Maintainability float64  `json:"maintainability,omitempty"`
	ChurnAbsolute   int      `json:"churnAbsolute"`
	ChurnRelative   float64  `json:"churnRelative"`
	ChangeFrequency int      `json:"changeFrequency"`
	HotspotScore    float64  `json:"hotspotScore"`
	AgeDays         int      `json:"ageDays"`
	Coverage        *float64 `json:"coverage,omitempty"`
	PrimaryAuthor   string   `json:"primaryAuthor,omitempty"`
	AuthorCount     int      `json:"authorCount,omitempty"`
}

// Coupling is a pairwise logical-coupling edge: files that repeatedly change together. Degree is
// symmetric (a same-commit co-change cannot indicate direction — research §3).
type Coupling struct {
	EntityA     string  `json:"entityA"`
	EntityB     string  `json:"entityB"`
	Degree      float64 `json:"degree"`
	SharedRevs  int     `json:"sharedRevisions"`
	AverageRevs float64 `json:"averageRevisions"`
}

// Ownership maps authors to their fractional ownership of a file and the file's bus factor (the
// fewest authors whose combined share exceeds 50%, OD-CI-3 default).
type Ownership struct {
	Path      string             `json:"path"`
	Authors   map[string]float64 `json:"authors"`
	BusFactor int                `json:"busFactor"`
}

// Summary holds the repo-level scalars and the A–E maintainability rating (contract §3).
type Summary struct {
	Lines                 int               `json:"lines"`
	EntityCount           int               `json:"entityCount"`
	TechnicalDebtRatio    float64           `json:"technicalDebtRatio"`
	MaintainabilityRating string            `json:"maintainabilityRating"`
	Coverage              *float64          `json:"coverage,omitempty"`
	BusFactor             int               `json:"busFactor"`
	Dora                  *DoraKeys         `json:"dora,omitempty"`
	Ratings               map[string]string `json:"ratings,omitempty"`
}

// DoraKeys is reserved (contract §4): populated only when a deploy/incident feed is wired, since
// deploy-frequency and change-failure-rate need data beyond the git repository.
type DoraKeys struct {
	DeploymentFrequency string  `json:"deploymentFrequency,omitempty"`
	LeadTimeHours       float64 `json:"leadTimeHours,omitempty"`
	ChangeFailureRate   float64 `json:"changeFailureRate,omitempty"`
	RecoveryHours       float64 `json:"recoveryHours,omitempty"`
	Performer           string  `json:"performer,omitempty"`
}

// TrendSeries is a commit-hash-stamped time series for the long-term monitoring lens.
type TrendSeries struct {
	Metric string       `json:"metric"`
	Points []TrendPoint `json:"points"`
}

// TrendPoint samples one metric at one commit (the commit hash makes the sample reproducible).
type TrendPoint struct {
	Commit string  `json:"commit"`
	At     string  `json:"at"`
	Value  float64 `json:"value"`
}

// View is one entry in the dynamic render-plan: which widget primitive renders which slice of the
// Report at which SDLC attention point (contract §5). The frontend iterates Views and instantiates
// the named primitive bound to the encoded fields — so adding a view is data, not code.
type View struct {
	ID             string            `json:"id"`
	Title          string            `json:"title"`
	Primitive      string            `json:"primitive"`
	AttentionPoint string            `json:"attentionPoint"`
	Encoding       map[string]string `json:"encoding"`
	Source         string            `json:"source"`
}

File: typescript/visualization/src/report/report.ts
/**
 * `@eden/visualization` — the codeinsight `Report` payload types (the CONSUMER half of the
 * abstraction seam). See docs/architecture/contracts/codeinsight.md §3 (Report schema) + §5 (Views
 * render-plan).
 *
 * `codeinsight` (the Go analyzer, the producer) walks a git repository and emits ONE self-describing
 * `Report`: per-entity static + behavioral metrics, pairwise logical coupling, repo-level ratings,
 * ownership, trends, and a `Views` render-plan. The `Report` JSON is the wire shape; these TypeScript
 * types mirror the Go struct's JSON tags EXACTLY so the analyzer (producer) and `@eden/visualization`
 * (consumer) build independently against the same contract — a frontend renders any repo's dashboard
 * with zero per-repo code (contract §1, §3).
 *
 * ONE CONCEPT, ONE HOME (10 §9): the wire shape is owned by the Go `codeinsight.Report`; this module
 * is the cited TypeScript projection of it, never a second source of truth. A producer field added at
 * the contract is mirrored here; this file does not invent fields the contract does not declare.
 *
 * Optionality maps the Go JSON tags: a `,omitempty` Go field is an OPTIONAL TypeScript property
 * (`field?:`) because the producer omits it when zero/absent; a non-omitempty field is required. A
 * Go pointer (`*float64`) that is `,omitempty` is `field?: number` (absent when nil). Numbers are Go
 * ints/float64 (both `number`); strings are RFC3339 / hex / free text per the contract comment.
 */

/** Where a `Report` was computed — the repository identity + the commit it was snapshotted at (§3). */
export interface RepositoryRef {
  /** Logical name (never a secret / a URL carrying credentials). */
  readonly identifier: string;
  /** The 40-hex commit the snapshot was taken at. */
  readonly headCommit: string;
  /** RFC3339; stamped by the caller's Clock. */
  readonly analyzedAt: string;
  readonly commitCount: number;
}

/** Which slice of history fed the temporal metrics — the analyzed commit range (§3). */
export interface Window {
  /** Lower commit bound (omitted → open-ended start). */
  readonly fromCommit?: string;
  /** Upper commit bound — usually {@link RepositoryRef.headCommit}. */
  readonly toCommit: string;
  /** RFC3339 lower bound (omitted when not time-bounded). */
  readonly since?: string;
  /** RFC3339 upper bound (omitted when not time-bounded). */
  readonly until?: string;
  /** Commits in the window. */
  readonly revisions: number;
}

/**
 * A per-file / per-package node — the unit the hotspot-map, enclosure, heatmap, and ownership-map
 * primitives render (contract §3). Every numeric datum is reproducible from `(repository, window)`.
 * The reserved (`omitempty`) fields fill as the analyzer's computation lands; this shape does not
 * change when they do (contract §4).
 */
export interface Entity {
  readonly path: string;
  /** `"file"` | `"directory"` | `"package"` (contract OD-CI-2: file is the v1 default). */
  readonly kind: string;
  readonly language?: string;
  readonly lines: number;
  readonly cyclomatic?: number;
  /** Cognitive complexity (§R2 tooling). */
  readonly cognitive?: number;
  /** 0–100, VS-rescaled. */
  readonly maintainability?: number;
  /** added + deleted across the window. */
  readonly churnAbsolute: number;
  /** churnAbsolute / lines — the preferred defect predictor (contract §3/§4). */
  readonly churnRelative: number;
  /** #commits touching this entity in the window. */
  readonly changeFrequency: number;
  /** changeFrequency × complexity, 0–1 normalized — the hotspot intensity (contract §3). */
  readonly hotspotScore: number;
  /** Days since the last substantive change. */
  readonly ageDays: number;
  /** Ingested from a coverage profile when present (Go `*float64`, omitempty). */
  readonly coverage?: number;
  readonly primaryAuthor?: string;
  readonly authorCount?: number;
}

/** A pairwise logical-coupling edge — feeds the dependency-matrix primitive (contract §3). */
export interface Coupling {
  readonly entityA: string;
  readonly entityB: string;
  /** 0–100 = sharedRevisions / averageRevisions. */
  readonly degree: number;
  readonly sharedRevisions: number;
  /** Gate ≥10 to suppress accidental co-change (contract §3). */
  readonly averageRevisions: number;
}

/** Author → code attribution + bus-factor inputs — feeds the ownership-map primitive (contract §3). */
export interface Ownership {
  readonly path: string;
  /** author → fractional line ownership (0–1). */
  readonly authors: Readonly<Record<string, number>>;
  /** #authors covering >50% (default; contract OD-CI-3). */
  readonly busFactor: number;
}

/** DORA keys — bands + derivation land from research round 2; mostly `omitempty` (contract §3/§4). */
export interface DoraKeys {
  readonly deploymentFrequency?: string;
  readonly leadTimeHours?: number;
  readonly changeFailureRate?: number;
  readonly recoveryHours?: number;
  /** Elite | High | Medium | Low. */
  readonly performer?: string;
}

/** Repo-level scalars + the A–E ratings — feeds the rating-badge primitive (contract §3). */
export interface Summary {
  readonly lines: number;
  readonly entityCount: number;
  /** SQALE; 0–1. */
  readonly technicalDebtRatio: number;
  /** `"A"`…`"E"` (≤5 / <10 / <20 / <50 / ≥50 %). */
  readonly maintainabilityRating: string;
  /** Repo-level coverage (Go `*float64`, omitempty). */
  readonly coverage?: number;
  /** Repo-level minimum critical set. */
  readonly busFactor: number;
  /** §R2 — present once a deploy/incident feed is wired (Go `*DoraKeys`, omitempty). */
  readonly dora?: DoraKeys;
  /** dimension → `"A".."E"` (omitempty). */
  readonly ratings?: Readonly<Record<string, string>>;
}

/** One sample of a time series — `value` taken at the commit it was computed at (contract §3). */
export interface TrendPoint {
  /** The commit hash this sample was taken at. */
  readonly commit: string;
  /** RFC3339. */
  readonly at: string;
  readonly value: number;
}

/** A time series for the monitoring lens — feeds the trend primitive (contract §3). */
export interface TrendSeries {
  /** `"churn"` | `"debtRatio"` | `"coverage"` | a dora key. */
  readonly metric: string;
  readonly points: readonly TrendPoint[];
}

/**
 * The producer's declaration of WHICH widget primitive renders WHICH slice of the report at WHICH
 * SDLC attention point (contract §5). The frontend iterates `Views` and, for each, instantiates the
 * named primitive bound to the encoded fields — so adding a view is DATA, not code.
 */
export interface View {
  readonly id: string;
  readonly title: string;
  /** One of the 8 widget primitives (contract §5) — e.g. `"hotspot-map"`. */
  readonly primitive: string;
  /** `"pre-commit"` | `"pull-request"` | `"release-gate"` | `"monitoring"`. */
  readonly attentionPoint: string;
  /**
   * channel → report field. e.g. the hotspot-map default:
   * `{ x: "churnRelative", y: "cyclomatic", size: "lines", color: "hotspotScore", label: "path" }`.
   */
  readonly encoding: Readonly<Record<string, string>>;
  /** Which Report collection feeds it: `"entities"|"couplings"|"ownership"|"trends"|"summary"`. */
  readonly source: string;
}

/**
 * The one self-describing envelope (contract §3). Every datum is stamped with the commit it was
 * computed at, so a `Report` is replayable and diff-able across history.
 */
export interface Report {
  /** semver of THIS contract. */
  readonly schemaVersion: string;
  readonly repository: RepositoryRef;
  /** The commit range analyzed. */
  readonly window: Window;
  /** Per-file / per-package nodes. */
  readonly entities: readonly Entity[];
  /** Pairwise logical-coupling edges. */
  readonly couplings: readonly Coupling[];
  /** author→code, bus-factor inputs. */
  readonly ownership: readonly Ownership[];
  /** Repo-level scalars + A–E ratings. */
  readonly summary: Summary;
  /** Time series for the monitoring lens. */
  readonly trends: readonly TrendSeries[];
  /** The dynamic render-plan (contract §5). */
  readonly views: readonly View[];
}
